torch/_export/serde/type_utils.py

```diff
@@ -0,0 +1,126 @@
+    """
+
+    head, args = deconstruct(t)
+    if head == list:
+        return f"List[{pp(args[0])}]"
+    elif head == set:
+        return f"Set[{pp(args[0])}]"
+    elif head == dict:
+        return f"Dict[{pp(args[0])}, {pp(args[1])}]"
+    elif head == Union:
+        args_str = ", ".join(pp(arg) for arg in args)
+        return f"Union[{args_str}]"
+    else:
+        return t.__name__
+
+
+def check(v: object, t: object) -> Union[Succeed, Fail, NotSure]:
+    """
+    Tries to decide if v is of type t
+
+    Succeeds
+        - when v is a simple Python value (e.g., ints, floats, nested built-in structs like Dict, Set, List, Union)
+          and v's type matches t
+        - or when `isinstance(v, t)`
+
+    Fails
+        - when v is a simple Python value (as described above)
+          and v's type does not match t
+
+    Gives NotSure
+        - when the type of v is not included in this logic
+    """
+
+    head, args = deconstruct(t)
+    if head == list:
+        if not isinstance(v, list):
+            return Fail(msg=[f"Expected {v} to be a list, but got {type(v).__name__}."])
+        element_type = args[0]
+        for element in v:
+            check_result = check(element, element_type)
+            if isinstance(check_result, Fail):
+                return Fail([f"{v} is not a {pp(t)}."] + check_result.msg)
+        return Succeed()
+    elif head == dict:
+        if not isinstance(v, dict):
+            return Fail(msg=[f"Expected {v} to be a dict, but got {type(v).__name__}."])
+        key_type, val_type = args[0], args[1]
+        for key, val in v.items():
+            check_result = check(key, key_type)
+            if isinstance(check_result, Fail):
+                return Fail([f"{v} is not a {pp(t)}."] + check_result.msg)
+            check_result = check(val, val_type)
+            if isinstance(check_result, Fail):
+                return Fail([f"{v} is not a {pp(t)}."] + check_result.msg)
+        return Succeed()
+    elif head == set:
+        if not isinstance(v, set):
+            return Fail(msg=[f"Expected {v} to be a set, but got {type(v).__name__}."])
+        element_type = args[0]
+        for element in v:
+            check_result = check(element, element_type)
+            if isinstance(check_result, Fail):
+                return Fail([f"{v} is not a {pp(t)}."] + check_result.msg)
+        return Succeed()
+    elif head == Union:
+        for arg in args:
+            check_result = check(v, arg)
+            if isinstance(check_result, Succeed):
+                return Succeed()
+        return Fail([f"{v} is not of type {pp(t)}."])
+    elif head in [int, float, str, bool] or inspect.isclass(head):
+        if isinstance(v, head):
+            return Succeed()
+        else:
+            return Fail([f"Expected {head.__name__} from {v}, but got {type(v).__name__}."])
+    else:
+        return NotSure(t=t, v=v)
+
+
+def deconstruct(t: object) -> Tuple[object, Tuple[object, ...]]:
```

Comment:
What happens if `t` is a custom class, like [`ExportedProgram`](https://github.com/pytorch/pytorch/blob/cebc11ae8fbc13de4eadee19c579712e5d282754/torch/_export/serde/schema.py#L222)? Do we have to add a case in `check` for every custom class?